src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -0,0 +1,250 @@
+                external_input.type.tensor_type.shape.dim[1].dim_value = sequence_length
+
+            elif external_input.name.startswith(_CACHE_INPUT_NAME):
+                # empty cache for multi-token runs,
+                # otherwise max cache len is max len - 1
+                external_input.type.tensor_type.shape.dim[2].dim_value = (
+                    0 if multitoken else sequence_length - 1
+                )
+            else:
+                raise ValueError(
+                    f"Unexpected external input name: {external_input.name}"
+                )
+            input_names.append(external_input.name)
+
+        _LOGGER.info(
+            f"Overwriting in-place the input shapes of the transformer model at {onnx_file_path}"
+        )
+        save_onnx(model, onnx_file_path)
+
+        is_cache_enabled = any(
+            _CACHE_INPUT_NAME in node.name for node in model.graph.initializer
+        )
+        return onnx_file_path, is_cache_enabled
+
+    def generate_token(self, logits: numpy.ndarray) -> numpy.ndarray:
+        """
+        Samples a token from the logits using the sampling temperature.
+        :param logits: the logits from the model with shape (vocab_size,)
+        :return: the sampled token
+        """
+        if self.deterministic:
+            return numpy.argmax(logits)
+
+        logits /= self.sampling_temperature
+
+        probs = softmax(logits)
+
+        return numpy.random.choice(len(probs), p=probs)
+
+    def _initialize_kv_cache_state(self, length: int) -> Dict[str, numpy.ndarray]:
+        # initialize empty kv cache of size
+        # (batch_size, num_attention_heads, length, hidden_dims)
+
+        cache_engine_input_index = next(
+            i
+            for i, name in enumerate(self.engine.input_names)
+            if _CACHE_INPUT_NAME in name
+        )
+        batch_size, num_attention_heads, _, hidden_dims = self.engine.input_shapes[
+            cache_engine_input_index
+        ]
+
+        empty_kv_cache_tensor = numpy.zeros(
+            (batch_size, num_attention_heads, length, hidden_dims),
+            dtype=numpy.float32,
+        )
+
+        cache_keys = [
+            output_name.replace("present", _CACHE_INPUT_NAME)
+            for output_name in self.engine.output_names
+            if output_name.startswith("present")
+        ]
+        return {key: empty_kv_cache_tensor for key in cache_keys}
+
+    def _add_kv_cache_to_input(self, inp: List[numpy.ndarray]) -> List[numpy.ndarray]:
+        # extend the engine input with the kv cache state
+        kv_cache_state = self.kv_cache.cached_inputs
+        if kv_cache_state is None:
+            # if kv cache state is None, we need to initialize the
+            # kv cache state and the session
+            kv_cache_state = self._initialize_kv_cache_state(
+                length=self.sequence_length if self.multitoken else 1
```

Comment:
len(input_ids) # 128 + input_cache_len # 0 = max_sequence_len # 128 = output_cash_shape_len # 128

# multitoken --> length = 0 # (64 in the future)
# single_token -> length = max_seq_len - 1